src/charm.py

```diff
@@ -1,233 +1,294 @@
+    @log_event_handler(logger)
+    def _on_update_status(self, event: UpdateStatusEvent) -> None:
+        """Handle `update-status` events.
 
         Args:
-            event:The event triggered by the restart action
+            event: The `update-status` event triggered at intervals.
         """
-        container = self.unit.get_container(self.name)
-        if not container.can_connect():
-            event.fail("Failed to connect to the container")
+        # TODO: Evaluate if this is needed
+        valid_pebble_plan = self._validate_pebble_plan()
+        if not valid_pebble_plan:
+            self._handle_status_update_config(event)
             return
 
-        self.unit.status = MaintenanceStatus("restarting application")
-        container.restart(self.name)
-
-        event.set_results({"result": "application successfully restarted"})
+        check = self._container.get_check("up")
+        if check.status != CheckStatus.UP:
+            self.unit.status = MaintenanceStatus("Status check: DOWN")
+            return
 
-    def _validate(self):
-        """Validate that configuration and relations are valid and ready.
+    @log_event_handler(logger)
+    def _on_oauth_info_changed(self, event: OAuthInfoChangedEvent) -> None:
+        """Handle `oauth-info-changed` event.
 
-        Raises:
-            ValueError: in case of invalid configuration.
+        Args:
+            event: The `OAuthInfoChangedEvent` event triggered when integration data changed.
         """
-        if not self.config["upstream"]:
-            raise ValueError("missing `upstream` config")
-
-        if not self.config["cookie-secret"]:
-            raise ValueError("missing `cookie-secret` config")
-
-        if self.config["provider"] == "google" and (not self.config["client-id"] or not self.config["client-secret"]):
-            raise ValueError("`client-id` and `client-secret` config must be set for `google` provider")
-
-        if not self.config["authenticated-emails-list"] and "--email-domain=" not in self.config["additional-config"]:
-            raise ValueError(
-                "`--email-domain` must be set using `additional-config` if not setting `authenticated-emails-list`"
-            )
+        self._handle_status_update_config(event)
 
     @log_event_handler(logger)
-    def _update(self, event):
-        """Update the application configuration and replan its execution.
+    def _handle_status_update_config(self, event: HookEvent) -> None:
+        """Update the application status and configuration and restart the container.
 
         Args:
-            event: The event triggered when the relation changed.
+            event: The event triggered when the application needs to be updated.
         """
-        try:
-            self._validate()
-        except ValueError as err:
-            self.unit.status = BlockedStatus(str(err))
+        if not self._container.can_connect():
+            event.defer()
```

Comment:
I always forget where we stand on this. Do we need the `defer` here? Since this is the holistic handler, this may be called by some hook when this changes. No problem with not handling this on this PR.

Reply:
Eventually it would be called by `update-status` hook if nothing else is emitted, but that all comes down to what its interval is. In the worst case the charm might end up unavailable till it runs, for that reason I still think that defer is a sensible solution here.
I might have missed sth, but if we chose to replace this with a pebble notice, it would be called whenever the container can be connected to - which would be happening repeatedly.

Reply:
No it's fine, let's leave it as it is for now